rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -32,261 +34,388 @@ trait HotStore[F[_], C, P, A, K] {
-      continuations <- getContinuations(join)
-      s             <- cacheRef.get
-      index         = s.joins(channel).indexOf(join)
-      _ <- if (continuations.isEmpty && index != -1)
-            checkIndex(s.joins(channel), index) >>
-              cacheRef.update { cache =>
-                val updated =
-                  removeIndex(
-                    cache.joins(
-                      channel
+      joinsInHistoryStore <- getJoinsFromHistoryStore(channel)
+      contsInHistoryStore <- getContFromHistoryStore(join)
+      err <- hotStoreState.modify[Boolean](state => {
+
+              val curJoins = (state.joins.get(channel) match {
+                // update with what is in historyStore and return the same
+                case None =>
+                  (
+                    state.copy(
+                      joins = state.joins.updated(channel, joinsInHistoryStore)
                     ),
-                    index
+                    joinsInHistoryStore
                   )
-                ignore(
-                  cache.joins.put(
-                    channel,
-                    updated
+                // just return what is in hot store already
+                case Some(joins) =>
+                  (state, joins)
+              })._2
+
+              val curConts = (state.continuations.get(join) match {
+                // update with what is in historyStore and return the same
+                case None =>
+                  (
+                    state.copy(
+                      continuations = state.continuations.updated(join, contsInHistoryStore)
+                    ),
+                    state.installedContinuations.get(join) ++: contsInHistoryStore
                   )
-                )
-                cache
-              } else ().pure[F]
+                // just return what is in hot store already
+                case Some(continuations) =>
+                  (state, state.installedContinuations.get(join) ++: continuations)
+              })._2
+
+              val index       = curJoins.indexOf(join)
+              val outOfBounds = !curJoins.isDefinedAt(index)
+
+              // TODO should attimpting to remove join with non empty contnuation lead to error as well?
+              val doRemove = curConts.isEmpty
+
+              if (doRemove) {
+                if (outOfBounds)
+                  (state.copy(joins = state.joins.updated(channel, curJoins)), true)
+                else {
+                  val newVal = removeIndex(curJoins, index)
+                  (state.copy(joins = state.joins.updated(channel, newVal)), false)
+                }
+              } else (state.copy(joins = state.joins.updated(channel, curJoins)), false)
+            })
+//      _ <- Sync[F]
+//            .raiseError(
+//              new IndexOutOfBoundsException(
+//                s"Index out of bounds when removing join"
+//              )
+//            )
+//            .whenA(err)
```

Comment:
The old implementation ignore error like this. Not sure if we should raise error for this. It looks we'd better raise it while some tests are failing because the original tests are too sticky on implementations.

Reply:
I think this error was there before in `checkIndex`? It's just checking of index has been done in the middle two state modifications that are supposed to be one atomic thing. So index check is performed inside pure computation and error is raised outside.

Reply:
Not sure what you mean here. But I made some changes to include it.

Reply:
I am still in a dilemma to do that.